dynamic-floating-group.lisp

```diff
@@ -0,0 +1,397 @@
+;; Copyright (C) Jin-Cheng Guu
+;;
+;;  This file is part of stumpwm.
+;;
+;; stumpwm is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 2, or (at your option)
+;; any later version.
+
+;; stumpwm is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this software; see the file COPYING.  If not, see
+;; <http://www.gnu.org/licenses/>.
+
+(in-package :stumpwm-dfg)
+
+(defparameter *default-layout* 'left-vertical)
+(defparameter *default-master-ratio* (/ 2 (+ 1 (sqrt 5))))
+
+(defstruct window+
+  "An augmented window (window+) is a window with some other
+information. Usually, in a dynamic floating group, the
+information will be used when the group re-tiles it."
+  ;; access example: (window+-free (make-window+ :window 1 :free t))
+  :window :free)
+
+;; A dyn-order, or a dynamic order, is a list of window+.
+(defclass dyn-float-group (stumpwm::float-group)
+  ((dyn-order :initform nil
+              :accessor dyn-float-group-dyn-order
+              :documentation
+              "The list of augmented windows (window+) that the
+              group holds. A dyn-float-group should tile
+              automatically according to it.")
+   (layout-hist :initform (list *default-layout*)
+                :accessor dyn-float-group-layout-hist
+                :documentation
+                "The list of layout histories. The first element
+                is interpreted as the current layout.")
+   (master-ratio :initform *default-master-ratio*
+                 :accessor dyn-float-group-master-ratio
+                 :documentation
+                 "The ratio that the master window will take
+                 apart.")))
+
+(defun dyn-float-group-p (group)
+  (eq (type-of group) 'dyn-float-group))
+
+(flet ((add-float-window (group window)
+         ;; not sure if needed
+         (change-class window 'stumpwm::float-window)
+         (stumpwm::float-window-align window)
+         (stumpwm::group-focus-window group window)))
```

Comment:
I'm not quite sure what you mean.. would you please elaborate a bit more?

Reply:
I'm saying that instead of repeating the same piece of code in two different local function definitions, lets define a top level function and use it on both places

Reply:
Got it! Done in [7e0fee4](https://github.com/stumpwm/stumpwm/pull/885/commits/7e0fee400e7c301b08371905ff65b0aceb6b2877).

+ Please note that this is the first time this branch changes the old code!
+ Please also note that `stumpwm:` will be taken away from `stumpwm:add-float-window` later.